man/man1/pmie2col.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMIE2COL 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmie2col\f1 \- convert pmie output to multi-column format
.SH SYNOPSIS
\f3pmie2col\f1
[\f3\-?\f1]
[\f3\-d\f1 \f2delimiter\f1]
[\f3\-p\f1 \f2precision\f1]
[\f3\-w\f1 \f2width\f1]
.SH DESCRIPTION
.de EX
.in +0.5i
.ie t .ft CB
.el .ft B
.ie t .sp .5v
.el .sp
.ta \\w' 'u*8
.nf
..
.de EE
.fi
.ie t .sp .5v
.el .sp
.ft R
.in
..
.B pmie2col
is a simple tool that converts output from
.BR pmie (1)
into regular column format.  Each column is 7 characters wide
(by default, may be changed with the
.B \-w
option) with a single space between columns.
That single space can be substituted with an alternate
delimiter using the
.B \-d
option (this is useful for importing the data into a spreadsheet,
for example).
.PP
The precision of the tabulated values from
.B pmie
can be specified with the
.B \-p
option (default is 2 decimal places).
This option can and will override any width setting in order to
present the requested precision.
.PP
The
.BR pmie (1)
configuration must follow these rules:
.IP (1)
Each
.BR pmie (1)
expression is of the form ``NAME = expr;''.
NAME will be used as the column heading, and must contain no white space,
although special characters can be escaped by enclosing NAME in single
quotes.
.IP (2)
The ``expr'' must be a valid
.BR pmie (1)
expression that produces a singular value.
.PP
In addition,
.BR pmie (1)
must be run with the
.B \-v
command line option.
.PP
It is also possible to use the
.B \-e
command line to
.BR pmie (1)
and output lines will be prefixed by a timestamp.
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-d\fR \fIchar\fR, \fB\-\-delimiter\fR=\fIchar\fR
Use \fIchar\fR as output delimiter.
.TP
\fB\-p\fR \fIN\fR, \fB\-\-precision\fR=\fIN\fR
Use \fIn\fR as output floating point precision.
.TP
\fB\-w\fR \fIN\fR, \fB\-\-width\fR=\fIN\fR
Use \fIn\fR as output column width.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH EXAMPLES
Given this
.BR pmie (1)
configuration file
.IR (config) :
.EX
loadav = kernel.all.load #'1 minute';
\&'%usr' = kernel.all.cpu.user;
\&'%sys' = kernel.all.cpu.sys;
\&'%wio' = kernel.all.cpu.wait.total;
\&'%idle' = kernel.all.cpu.idle;
\&'max-iops' = max_inst(disk.dev.total);
.EE
Then this command pipeline:
.EX
$ pmie \-v \-t 5 <config | pmie2col \-w 8
.EE
Produces output like this:
.EX
   loadav     %usr     %sys     %wio    %idle max-iops
     0.21        ?        ?        ?        ?        ?
     0.36     0.49     0.03     0.18     0.29    25.40
     0.49     0.41     0.10     0.36     0.13    51.00
     0.69     0.49     0.10     0.05     0.37    43.20
     0.71     0.39     0.08     0.04     0.49    14.00
     0.83     0.63     0.15     0.00     0.21    32.30
     1.09     0.60     0.02     0.10     0.27    47.00
     0.92     0.01     0.00     0.00     0.99     2.40
.EE
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1)
and
.BR pmie (1).

.\" control lines for scripts/man-spell
.\" +ok+ col {from pmie2col}
.\" +ok+ iops loadav sys wio {from column headings}
.\" +ok+ max_inst sp {.sp from troff macro}


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pmie/pmie2col.sh
#!/bin/sh
#
# A crude ascii reporting tool, convert pmie to column output
#
# The pmie rules need to be of the form:
# load_1 = kernel.all.load #'1 minute';
# idle = kernel.all.cpu.idle;
# column_name=some other expression;
# ...
#
# Each pmie expression has to produce a singular value.
#
# With timestamps (pmie -e or pmie output from a PCP archive), lines look like
#	metric (Tue Feb 13 05:01:19 2001): value
#       load_1 (Tue Dec 23 12:20:45 2003): 0.24
# the first sed step in the filter sorts this out.
#
# First e-mailed to Patrick Aland <paland@stetson.edu> and pcp@groups.io
# on Wed, 24 Jan 2001.
#
 
# Get standard environment
. $PCP_DIR/etc/pcp.env

tmp=`mktemp -d "$PCP_TMPFILE_DIR/pmie2col.XXXXXXXXX"` || exit 1
status=1
trap "rm -rf $tmp; exit \$status" 0 1 2 3 15
prog=`basename $0`

# For interactive use, works better with line buffered output from sed(1)
# and awk(1).
#
case "$PCP_PLATFORM"
in
    linux|mingw|kfreebsd|gnu|openbsd)
	SED="sed -u"
	;;
    freebsd|darwin)
	SED="sed -l"
	;;
    *)
	SED=sed
	;;
esac

echo > $tmp/usage
cat >> $tmp/usage <<EOF
Options:
  -d=CHAR,--delimiter=CHAR   set the output delimiter character
  -p=N,--precision=N   set the output floating point precision
  -w=N,--width=N       set the width of each column of output
  --help
EOF

_usage()
{
    pmgetopt --progname=$prog --config=$tmp/usage --usage
    exit 1
}

pre=2	# floating point precision
wid=7	# reporting column width
delim=' '

ARGS=`pmgetopt --progname=$prog --config=$tmp/usage -- "$@"`
[ $? != 0 ] && exit 1

eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1"
    in
	-w)	wid="$2"
		shift
		;;
	-p)	pre="$2"
		shift
		;;
	-d)	delim="$2"
		shift
		;;
	--)	shift
		break
		;;
	-\?)	_usage
		;;
    esac
    shift
done

[ $# -eq 0 ] || _usage

# culled output at the start is produced by pmie and/or pmafm
$SED \
    -e '/^pmie: timezone set to/d' \
    -e '/^Note: running pmie serially, once per archive$/d' \
    -e '/^Host: [A-Za-z0-9]* Archive: /d' \
    -e '/^[^ ][^ ]* ([A-Z][a-z][a-z] [A-Z][a-z][a-z]  *[0-9][0-9]* [0-2][0-9]:[0-5][0-9]:[0-5][0-9] [0-9][0-9][0-9][0-9]): /{
s/ (/|/
s/): /|/
}' \
    -e '/^\([^ ][^ ]*\):/s//\1||/' \
| awk -F\| -v delim="$delim" '
NF == 0				{ if (state == 0) {
				    ncol = i
				    print ""
				  }
				  if (stamp != "") printf "%24s ",stamp
				  for (i = 0; i < ncol; i++) {
				    if (v[i] == "?")
					# no value
					printf "%s%'$wid'.'$wid's",delim,"?"
				    else if (v[i]+0 == v[i])
					# number
					printf "%s%'$wid'.'$pre'f",delim,v[i]
				    else
					# string
					printf "%s%'$wid'.'$wid's",delim,v[i]
				    v[i] = "?"
				  }
				  print ""
				  i = 0
				  stamp = ""
				  state = 1
				  next
				}
NF == 3 && stamp == ""		{ stamp = $2 }
state == 0			{ if (i == 0 && stamp != "") printf "%24s ",""
				  printf "%s%'$wid'.'$wid's",delim,$1 }
				{ v[i++] = $NF }'

status=0
exit
